[Merged by Bors] - feat(Combinatorics): any connected graph contains a vertex that leaves the graph preconnected if removed

LLaurance opened the pull request:
This shows that any nontrivial connected graph contains a vertex that, when removed, leaves the graph connected.

---
- [x] depends on: #30590
- [x] depends on: #31103
- [x] depends on: #31116
- [x] depends on : #31118
- [x] depends on : #31119
- [x] depends on : #31120

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

LLaurance committed:
Nontrivial preconnected SimpleGraph has vertices of positive degree

LLaurance committed:
Nontrivial tree has minimum degree one

LLaurance committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into tree-nontrivial-mindegree

LLaurance committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into tree-nontrivial-mindegree

LLaurance committed:
Remove unused import

LLaurance committed:
Complete proof without bound tactic

LLaurance committed:
Update Mathlib/Combinatorics/SimpleGraph/Acyclic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

LLaurance committed:
Fix proof after suggested edit

LLaurance committed:
Fix proof after suggested edit

LLaurance committed:
Remove name from instance argument as suggested

LLaurance committed:
Improve readability

LLaurance committed:
snd_mem_tail_support

LLaurance committed:
Connected.connected_deleteVerts_singleton_of_degree_eq_one_of_nontrivial and supporting lemmas

LLaurance committed:
Minor improvements

LLaurance committed:
lemma _root_.SimpleGraph.toSubgraph.degree

LLaurance committed:
theorem Connected.toSubgraph

LLaurance committed:
lemma Connected_coeSubgraph

LLaurance committed:
lemma Connected.exists_vertex_connected_deleteVerts_singleton_of_nontrivial

LLaurance committed:
Minimize imports

LLaurance committed:
Merge branch 'tree-nontrivial-mindegree' of https://github.com/LLaurance/mathlib4 into connected-delete-vertex

LLaurance committed:
Replace non-terminal simps

LLaurance committed:
Use rw instead of simp only

LLaurance committed:
Update Acyclic.lean

Adjustments according to review comments by vlad902

LLaurance committed:
Merge branch 'tree-nontrivial-mindegree' of https://github.com/LLaurance/mathlib4 into connected-delete-vertex

LLaurance committed:
Replace non-terminal simps

LLaurance committed:
Style and minor changes

LLaurance committed:
Merge branch 'leanprover-community:master' into connected-delete-vertex

LLaurance committed:
Keep terminal simp_all

LLaurance committed:
Docstrings

LLaurance committed:
Minor change

LLaurance committed:
Shorten proofs

LLaurance committed:
Style

LLaurance committed:
Avoid using linarith since Combinatorics/SimpleGraph/Matching has assert_not_exists Field

LLaurance committed:
Shorten proof

LLaurance committed:
Merge branch 'master' into connected-delete-vertex

YaelDillies reviewed (commented):
Thank you for the PR!

You are proving a lot of lemmas about `Connected`. Would you please deduce them from analogous lemmas about `Preconnected`?

LLaurance committed:
SimpleGraph.toSubgraph.card_neighborSet_eq_degree

LLaurance committed:
SimpleGraph.Walk.getVert_mem_tail_support

LLaurance committed:
Change i > 0 to 0 < i

As suggested https://github.com/leanprover-community/mathlib4/pull/29309#discussion_r2422723996

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

LLaurance committed:
Revert change to theorem statement of Connected.coeSubgraph

LLaurance committed:
Rename theorem

According to suggestion https://github.com/leanprover-community/mathlib4/pull/29309/files#r2422708648

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

LLaurance committed:
Add protected modifier to theorem Connected.toSubgraph

According to suggestion https://github.com/leanprover-community/mathlib4/pull/29309/files#r2422711066

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

LLaurance committed:
Add protected modifier to lemma Connected.coeSubgraph

According to suggestion https://github.com/leanprover-community/mathlib4/pull/29309/files#r2422711159

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

LLaurance committed:
Replace appearances of Connected_coeSubgraph with new name Connected.coeSubgraph

LLaurance committed:
Write {x | x ∈ p.support} instead of p.toSubgraph.verts

LLaurance committed:
SimpleGraph.Subgraph.Connected.map_Subgraph_coe

LLaurance committed:
Remove need for Nontrivial instance

.Connectivity.Subgraph.Connected.connected_deleteVerts_singleton_of_degree_eq_one_of_nontrivial

LLaurance committed:
Rename theorem since no Nontrivial instance is required

`.Connectivity.Subgraph.Connected.connected_deleteVerts_singleton_of_degree_eq_one_of_nontrivial` to `.Connectivity.Subgraph.Connected.connected_deleteVerts_singleton_of_degree_eq_one`

LLaurance committed:
Relax theorem arguments

LLaurance committed:
def Walk.induce

LLaurance committed:
Helper theorems for SimpleGraph corresponding to existing ones for Subgraph

LLaurance committed:
lemma Preconnected.connected_induce_complement_singleton_of_degree_eq_one

SnirBroshi commented on Mathlib/Combinatorics/SimpleGraph/Walk.lean:
```diff
@@ -1377,6 +1377,12 @@ theorem reverse_transfer (hp) :
 
+def Walk.induce {G : SimpleGraph V} {u v : V} (s : Set V) :
+    (p : G.Walk u v) → (h : ∀ w, w ∈ p.support → w ∈ s) →
+    (G.induce s).Walk ⟨u, h _ (p.start_mem_support)⟩ ⟨v, h _ (p.end_mem_support)⟩
```
What's the advantage of a new recursive definition over using `SimpleGraph.Walk.transfer` to define it?

LLaurance replied:
> What's the advantage of a new recursive definition over using `SimpleGraph.Walk.transfer` to define it?

This immediately allows mapping a walk to a subgraph like `SimpleGraph ↑{v}ᶜ` by specifying the vertices.

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Finite.lean:
```diff
@@ -574,4 +574,15 @@ theorem card_edgeFinset_map (f : V ↪ W) (G : SimpleGraph V) [DecidableRel G.Ad
+    (h : G.degree v ≠ 0) : Nontrivial V := by
+  rw [Nat.ne_zero_iff_zero_lt, degree_pos_iff_exists_adj] at h
+  obtain ⟨w, hw⟩ := h
+  exact nontrivial_of_ne v w hw.ne
```
Can you prove this in terms of a new lemma `degree_eq_zero_of_subsingleton`?

LLaurance committed:
Walk.induce docstring

LLaurance committed:
theorem degree_eq_zero_of_subsingleton

+ new proof of theorem nontrivial_of_degree_ne_zero

LLaurance committed:
Amend docstring

LLaurance committed:
Amend docstring, change Fintype V to Fintype H.verts

LLaurance committed:
Shorten proof

LLaurance committed:
Revert "Shorten proof"

This reverts commit 720762c39ead729389385e5e53d0ff91db16014a.

LLaurance committed:
Shorten proof

Considering using `classical` in the proof but it is unclear to me the conventions and general preferences of using it

LLaurance committed:
Corresponding theorems for SimpleGraph

YaelDillies commented:
I also note that you deleted almost all of the default PR description. This is mostly okay, except for two things that you really need to keep:
* the `---` separator, which is what bors uses to pick out the git commit description (which you removed here)
* the big Ona button (which you didn't remove, thanks)

LLaurance committed:
Merge branch 'leanprover-community:master' into connected-delete-vertex

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:
```diff
@@ -797,11 +797,36 @@ theorem degree_spanningCoe {G' : G.Subgraph} (v : V) [Fintype (G'.neighborSet v)
   simp only [Set.mem_toFinset, mem_neighborSet]
 
+theorem nontrivial_of_degree_ne_zero {G' : Subgraph G} {v : V} [Fintype (G'.neighborSet v)]
+    (h : G'.degree v ≠ 0) : Nontrivial G'.verts := by
```
You're still missing `degree_eq_zero_of_subsingleton` here

LLaurance committed:
theorem degree_eq_zero_if_notMem_verts

LLaurance committed:
Subgraph version of theorem degree_eq_zero_of_subsingleton

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:
```diff
@@ -776,6 +776,13 @@ theorem finset_card_neighborSet_eq_degree {G' : Subgraph G} {v : V} [Fintype (G'
   rw [degree, Set.toFinset_card]
 
+theorem degree_eq_zero_if_notMem_verts {G' : Subgraph G} {v : V} [Fintype (G'.neighborSet v)]
+    (h : v ∉ G'.verts) : G'.degree v = 0 := by
```
```suggestion
theorem degree_of_notMem_verts {G' : Subgraph G} {v : V} [Fintype (G'.neighborSet v)]
    (h : v ∉ G'.verts) : G'.degree v = 0 := by
```
is probably enough

LLaurance committed:
Change appearances of renamed theorem

LLaurance committed:
Informative comments on the proof

LLaurance committed:
Informative comments on the proof

LLaurance committed:
Rename theorem

As suggested https://github.com/leanprover-community/mathlib4/pull/29309#discussion_r2436711197

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

LLaurance committed:
Change theorem degree_eq_zero_of_subsingleton parameters

As suggested https://github.com/leanprover-community/mathlib4/pull/29309#discussion_r2436712193

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

LLaurance committed:
Amend theorem parameters

LLaurance committed:
Update Acyclic.lean

LLaurance committed:
Rename Preconnected parameter

LLaurance committed:
Rename parameter

LLaurance committed:
Subgraph version of preconnected_bot

LLaurance committed:
Subgraph version of lemma Connected.exists_vertex_preconnected_deleteVerts_singleton_of_fintype

LLaurance committed:
Update Acyclic.lean

LLaurance committed:
Subgraph.coeInduceEquiv

LLaurance committed:
Shorten proof of Preconnected.connected_deleteVerts_singleton_of_degree_eq_one

Use the SimpleGraph version to prove the SimpleGraph.Subgraph version

LLaurance committed:
Update Subgraph.lean

LLaurance committed:
Shorten proof

LLaurance committed:
Shorten proof of lemma Preconnected.exists_vertex_connected_deleteVerts_singleton_of_fintype_of_nontrivial

LLaurance committed:
Shorten proof of lemma Connected.exists_vertex_preconnected_deleteVerts_singleton_of_fintype

LLaurance committed:
def coeDeleteVertsEquiv

LLaurance committed:
Use coeDeleteVertsEquiv in place of coeInduceEquiv

LLaurance committed:
Update Acyclic.lean

YaelDillies commented:
Can you merge master?

LLaurance committed:
Merge branch 'master' into connected-delete-vertex

LLaurance committed:
Merge branch 'connected-delete-vertex' of https://github.com/LLaurance/mathlib4 into connected-delete-vertex

LLaurance committed:
Delete my Walk.induce

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Finite.lean:
For example, this looks like a self contained PR

YaelDillies reviewed (commented):
Here's the PR from ProofBench: #31103

LLaurance committed:
Style

LLaurance committed:
Rearrange theorems to be consistent with #31116

LLaurance committed:
Remove unnecessary theorem preconnected_bot

LLaurance committed:
Merge branch 'master' into connected-delete-vertex

LLaurance committed:
fix proof after merge

LLaurance committed:
fix proof after merge

LLaurance committed:
Merge branch 'master' into connected-delete-vertex

LLaurance committed:
Merge branch 'master' into connected-delete-vertex

LLaurance committed:
Merge branch 'master' into connected-delete-vertex

LLaurance committed:
Merge branch 'master' into connected-delete-vertex

LLaurance committed:
Merge branch 'master' into connected-delete-vertex

LLaurance committed:
merge

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:
```diff
@@ -1231,6 +1231,15 @@ instance instDecidableRel_induce_adj (s : Set V) [∀ a, Decidable (a ∈ s)] [D
   fun _ _ ↦ instDecidableAnd
 
+/-- Equivalence between an induced Subgraph and its corresponding SimpleGraph. -/
+def coeInduceEquiv {s : Set V} (h : s ⊆ G'.verts) :
```
```suggestion
def coeInduceIso (s : Set V) (h : s ⊆ G'.verts) :
```

LLaurance replied:
Resolved https://github.com/leanprover-community/mathlib4/pull/29309/commits/5faf811a44ec45ba80baf8e3ec17acd5b1b42e39

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -333,4 +333,65 @@ lemma IsTree.exists_vert_degree_one_of_nontrivial [Fintype V] [Nontrivial V] [De
+/-- The graph resulting from removing a vertex of degree one from a (pre)connected graph is
+connected. -/
+lemma Preconnected.connected_induce_complement_singleton_of_degree_eq_one [DecidableEq V]
+    (hpreconn : G.Preconnected) {v : V} [Fintype ↑(G.neighborSet v)] (hdeg : G.degree v = 1) :
```
We only use `Preconnected` when the lemma could be useful to deal with the empty graph which isn't the case here since you have `v : V`
```suggestion
lemma Connected.induce_compl_singleton_of_degree_eq_one [DecidableEq V]
    (hconn : G.Connected) {v : V} [Fintype ↑(G.neighborSet v)] (hdeg : G.degree v = 1) :
```

LLaurance replied:
Resolved https://github.com/leanprover-community/mathlib4/commit/f94b76d566bcefba58e1a646565ce4c2637d2379
I neglected to split the changes into smaller commits, sorry.

LLaurance committed:
Revert import change

LLaurance committed:
Rename lemmas and tidy up

LLaurance commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Subgraph.lean:
```diff
@@ -5,8 +5,7 @@ Authors: Kyle Miller, Rémi Bottinelli
 
-public import Mathlib.Combinatorics.SimpleGraph.Connectivity.Connected
-public import Mathlib.Data.Set.Card
+public import Mathlib.Combinatorics.SimpleGraph.Acyclic
```
No.

LLaurance committed:
Rename to coeDeleteVertsIso

LLaurance committed:
Rename to coeInduceIso

LLaurance committed:
Update Acyclic.lean

LLaurance commented on Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:
```diff
@@ -1295,6 +1304,15 @@ instance instDecidableRel_deleteVerts_adj (u : Set V) [r : DecidableRel G.Adj] :
 
+/-- Equivalence between a Subgraph with deleted vertices and its corresponding SimpleGraph. -/
+def coeDeleteVertsEquiv {s : Set V} :
+    (G'.deleteVerts s).coe ≃g G'.coe.induce {v : G'.verts | ↑v ∈ s}ᶜ where
```
Resolved https://github.com/leanprover-community/mathlib4/pull/29309/commits/7970151cee5fd70fbcdc36b11c8bcfd1539dda73

LLaurance commented on Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:
```diff
@@ -1231,6 +1231,15 @@ instance instDecidableRel_induce_adj (s : Set V) [∀ a, Decidable (a ∈ s)] [D
+  toFun := fun ⟨v, hv⟩ ↦ ⟨⟨v, h hv⟩, by simp at hv; aesop⟩
+  invFun := fun ⟨v, hv⟩ ↦ ⟨v, hv⟩
+  left_inv := fun v ↦ by simp
+  right_inv := fun v ↦ by simp
```
Resolved https://github.com/leanprover-community/mathlib4/pull/29309/commits/5faf811a44ec45ba80baf8e3ec17acd5b1b42e39

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -333,4 +333,61 @@ lemma IsTree.exists_vert_degree_one_of_nontrivial [Fintype V] [Nontrivial V] [De
+lemma Connected.induce_compl_singleton_of_degree_eq_one (hconn : G.Connected) {v : V}
+    [Fintype ↑(G.neighborSet v)] (hdeg : G.degree v = 1) : (G.induce {v}ᶜ).Connected := by
+  obtain ⟨u, adj_vu, hu⟩ := degree_eq_one_iff_existsUnique_adj.mp hdeg
+  refine (connected_iff _).mpr ⟨?_, ?_⟩
```
```suggestion
  refine (connected_iff _).mpr ⟨?_, u, by aesop⟩
```
then you can unindent the rest

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -333,4 +333,61 @@ lemma IsTree.exists_vert_degree_one_of_nontrivial [Fintype V] [Nontrivial V] [De
+      ext w x
+      simp [subsingleton_iff.mp _ w x]
+    rw [this]
+    exact preconnected_bot
```
There should be a lemma saying that a graph on at most one vertex is preconnected

LLaurance replied:
I was unable to find such a lemma, but it can be added: https://github.com/leanprover-community/mathlib4/pull/29309/commits/b9a589ff49a1e716b73968fbb75d677dc19f5565

YaelDillies replied:
Yes sorry, I meant you should write it!

LLaurance committed:
lemma reachable_of_subsingleton

LLaurance committed:
lemma preconnected_of_subsingleton

LLaurance committed:
Update Acyclic.lean

LLaurance committed:
Update Connected.lean

LLaurance committed:
lemma Connected.of_nonempty_subsingleton

LLaurance committed:
Rename occurrence of renamed lemma

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -156,6 +156,9 @@ lemma reachable_bot {u v : V} : (⊥ : SimpleGraph V).Reachable u v ↔ u = v :=
   · simp
   · exact ⟨.cons huv .nil⟩
 
+lemma reachable_of_subsingleton [Subsingleton V] {u v : V} : G.Reachable u v := by
```
```suggestion
lemma Reachable.of_subsingleton [Subsingleton V] {u v : V} : G.Reachable u v := by
```

LLaurance committed:
Adjust arguments

LLaurance committed:
Update Acyclic.lean

LLaurance committed:
Update Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

LLaurance committed:
Update Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -156,6 +156,9 @@ lemma reachable_bot {u v : V} : (⊥ : SimpleGraph V).Reachable u v ↔ u = v :=
   · simp
   · exact ⟨.cons huv .nil⟩
 
+lemma Reachable.of_subsingleton [Subsingleton V] {u v : V} : G.Reachable u v := by
```
I would make `G` implicit here (and similarly for the other two)

LLaurance replied:
Resolved https://github.com/leanprover-community/mathlib4/pull/29309/commits/30a68a043586490d4b12040b4e1537c2f33689c1

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -190,6 +193,9 @@ alias bot_preconnected_iff_subsingleton := preconnected_bot_iff_subsingleton
 @[deprecated (since := "2025-09-23")] alias top_preconnected := preconnected_top
 
+lemma Preconnected.of_subsingleton [Subsingleton V] (G : SimpleGraph V) : G.Preconnected :=
+  fun _ _ ↦ Reachable.of_subsingleton G
```
```suggestion
  fun _ _ ↦ .of_subsingleton
```

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -301,6 +307,10 @@ lemma connected_top_iff : (completeGraph V).Connected ↔ Nonempty V := by simp
 @[deprecated (since := "2025-09-23")] alias bot_not_connected := not_connected_bot
 @[deprecated (since := "2025-09-23")] alias top_connected := connected_top
 
+lemma Connected.of_nonempty_subsingleton [Nonempty V] [Subsingleton V] (G : SimpleGraph V) :
```
```suggestion
lemma Connected.of_subsingleton [Nonempty V] [Subsingleton V] (G : SimpleGraph V) :
```
is an acceptable lie, I would think, to preserve the series of lemmas with the same names

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean:
```diff
@@ -301,6 +307,10 @@ lemma connected_top_iff : (completeGraph V).Connected ↔ Nonempty V := by simp
 
+lemma Connected.of_nonempty_subsingleton [Nonempty V] [Subsingleton V] (G : SimpleGraph V) :
+    G.Connected :=
+  ⟨Preconnected.of_subsingleton G⟩
```
```suggestion
  ⟨.of_subsingleton⟩
```

LLaurance committed:
Update Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

LLaurance committed:
Update Mathlib/Combinatorics/SimpleGraph/Connectivity/Connected.lean

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

LLaurance committed:
Tag lemmas with @[nontriviality]

LLaurance commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -333,4 +333,55 @@ lemma IsTree.exists_vert_degree_one_of_nontrivial [Fintype V] [Nontrivial V] [De
+    exact ⟨v, hv.preconnected⟩
+  · use hconn.nonempty.some
+    have := not_nontrivial_iff_subsingleton.mp h
+    exact Preconnected.of_subsingleton _
```
I cannot make it work :( I can tag the new lemmas with `@[nontriviality]` : https://github.com/leanprover-community/mathlib4/pull/29309/commits/568cd0ad689ed8c3b40b312bdf71d1d4e34c9c3f

LLaurance committed:
Update docComment

LLaurance committed:
Update Acyclic.lean

LLaurance committed:
Rename lemma

LLaurance committed:
Update Mathlib/Combinatorics/SimpleGraph/Subgraph.lean

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

LLaurance committed:
use nontriviality

LLaurance commented on Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:
```diff
@@ -1231,6 +1231,13 @@ instance instDecidableRel_induce_adj (s : Set V) [∀ a, Decidable (a ∈ s)] [D
     DecidableRel (G'.induce s).Adj :=
   fun _ _ ↦ instDecidableAnd
 
+/-- Equivalence between an induced Subgraph and its corresponding SimpleGraph. -/
```
Resolved https://github.com/leanprover-community/mathlib4/pull/29309/commits/bd9572d3df2a75d9f6c5918f3fd0c420b61a9f1b https://github.com/leanprover-community/mathlib4/pull/29309/commits/0bbdd6114a50085d5aebd5312e3a761eb25d7495

LLaurance committed:
Update Mathlib/Combinatorics/SimpleGraph/Acyclic.lean

Co-authored-by: Yaël Yanis Dillies <yael.dillies@gmail.com>

LLaurance commented on Mathlib/Combinatorics/SimpleGraph/Acyclic.lean:
```diff
@@ -333,4 +333,55 @@ lemma IsTree.exists_vert_degree_one_of_nontrivial [Fintype V] [Nontrivial V] [De
+    exact ⟨v, hv.preconnected⟩
+  · use hconn.nonempty.some
+    have := not_nontrivial_iff_subsingleton.mp h
+    exact Preconnected.of_subsingleton
```
That part can be `simp [hconn.nonempty]` (regardless of whether any of the new lemmas are tagged with `@[simp]`) or what it is now `exact ⟨hconn.nonempty.some, .of_subsingleton⟩`

YaelDillies reviewed (approved):
Thanks! 🚀

maintainer merge

jcommelin reviewed (commented):
Thanks :tada:

bors merge
